src/9FC00B40.c

```diff
@@ -39,27 +117,387 @@ s32 skGetId(BbId* id) {
+        : : "r" (a0), "i" (launch_app_trampoline) : "v0", "t0"
+    );
+
+    return -1;
+}
+
+s32 skRecryptListValid(RecryptList* recryptList) {
+    if (!CHECK_UNTRUSTED_RECRYPT_LIST(recryptList)) {
+        return -1;
+    }
+
+    return recrypt_list_verify_size_and_sig(recryptList);
+}
+
+s32 skRecryptBegin(BbTicketBundle* bundle, BbAppLaunchCrls* crls, RecryptList* recryptList) {
+    BbAesKey sp10;
+    BbTicketHead* head;
+    s32 ret;
+
+    head = &bundle->ticket->head;
+    ret = func_9FC00DCC(bundle, crls, recryptList, TRUE);
+
+    if (ret) {
+        return ret;
+    }
+
+    ret = recrypt_list_get_key_for_cid(recryptList, &sp10, contentMetaDataHead.id);
+
+    if (ret == 3) {
+        D_9FC0F2DC = TRUE;
+        aes_cbc_set_key_iv(&sp10, &contentMetaDataHead.iv);
+    } else {
+        D_9FC0F2DC = FALSE;
+        aes_cbc_set_key_iv(&contentMetaDataHead.key, &contentMetaDataHead.iv);
+        aesCipherInit(&D_9FC0F2C8, 2, &contentMetaDataHead.iv);
+        recrypt_list_add_new_entry(recryptList, contentMetaDataHead.id, 3);
+    }
+
+    aesMakeKey(&D_9FC0F0E0, 0, 0x80, &sp10);
+    SHA1Reset(&sha1_ctx);
+    memcpy(&ticketHead, head, sizeof(BbTicketHead));
+
+    D_9FC0F0DC = 0;
+    D_9FC0F2E0 = FALSE;
+
+    return ret;
+}
+
+s32 func_9FC0134C(u8* buf, u32 size, s32 a2) {
+    u32 chunkSize = 0x200;
+    u32 left;
+    u32 i;
+
+    for (i = 0; i < size; i += chunkSize) {
+        if (dma_from_cart(FALSE, K0_TO_PHYS(buf), chunkSize, TRUE) < 0) {
+            return -1;
+        }
+
+        func_9FC0384C(0, D_9FC0F2E0);
+        D_9FC0F2E0 = TRUE;
+
+        while (IO_READ(PI_AES_STATUS_REG) & PI_AES_BUSY);
+
+        left = contentMetaDataHead.size - D_9FC0F0DC;
+
+        if (left >= chunkSize) {
+            SHA1Input(&sha1_ctx, PHYS_TO_K1(PI_10000_BUF_START), chunkSize);
+            D_9FC0F0DC += chunkSize;
+        } else {
+            SHA1Input(&sha1_ctx, PHYS_TO_K1(PI_10000_BUF_START), left);
+            D_9FC0F0DC = contentMetaDataHead.size;
+        }
+
+        if (a2) {
+            aesBlockEncrypt(&D_9FC0F2C8, &D_9FC0F0E0, PHYS_TO_K1(PI_10000_BUF_START), chunkSize * 8, buf);
+        }
+
+        buf += chunkSize;
+    }
+
+    return 0;
+}
+
+s32 skRecryptData(u8* buf, u32 size) {
+    BbAesIv* iv1;
+    BbAesIv* iv2;
+
+    if (buf != NULL && !CHECK_UNTRUSTED_ARRAY(buf, size)) {
+        return -1;
+    }
+
+    if (D_9FC0F2DC == TRUE) {
+        iv1 = &D_9FC0F2E4;
+        iv2 = (BbAesIv*) (buf + size - sizeof(BbAesIv));
+
+        if (buf == NULL) {
+            iv1 = &contentMetaDataHead.iv;
+            iv2 = &contentMetaDataHead.iv;
+        }
+
+        aesCipherInit(&D_9FC0F2C8, 2, iv1);
+        aes_cbc_set_key_iv(contentMetaDataHead.key, iv2);
+
+        D_9FC0F2E0 = FALSE;
+        D_9FC0F2DC = FALSE;
+
+        return 0;
+    }
+
+    func_9FC0134C(buf, size, TRUE);
+
+    return 0;
+}
+
+s32 skRecryptComputeState(u8* buf, u32 size) {
+    BbAesIv* src = (BbAesIv*) (buf + size - sizeof(BbAesIv));
```

Comment:
perhaps something like `(BbAesIv*)&buf[size - sizeof(BbAesIv)]`? Not sure if I like it anymore